### ISSUE TITLE
Problematic macros in dart/collision/fcl/tri_tri_intersection_test.hpp

### ISSUE THREAD

**Issue by scpeters** [dartsim/dart#1186]
> If you are reporting a bug, please use the following template.

---

### Bug Report
> Please answer the following questions for yourself before submitting an issue.
- [X] I checked the [documentation](http://dartsim.github.io/) and found no answer.
- [X] I checked to make sure that this issue has not already been filed.

#### Environment
* Select DART version: 6.6.1
* Select OS name and version name(or number): macOS High Sierra
* Select compiler name and version number: Clang from Xcode 10

#### Expected Behavior
> Please describe the behavior you are expecting.

Include `dart/dart.hpp` in some code without having macro collisions with other library code.

#### Current Behavior
> What is the current behavior?

When attempting to compile [gazebo pull request 2956 (DART heightmaps)](https://bitbucket.org/osrf/gazebo/pull-requests/2956/dart-heightmap-with-bullet-and-ode/diff) on macOS (reproduced on both Sierra and High Sierra), I get a [compiler error](https://bitbucket.org/osrf/gazebo/pull-requests/2956/dart-heightmap-with-bullet-and-ode/diff#comment-78382703) because MacTypes.h has a `#define nil` that conflicts with a protected variable `nil` somewhere in an fcl class. That MacTypes.h macro is really dumb and not your fault.

As a workaround, I put the `#include <dart/dart.hpp>` first, so `MacTypes.h` wouldn't be able to mess it up. This sadly led to a conflict with Ogre, which indescribably has a class with protected variables in all caps that conflicts with a macro defined in `dart/collision/fcl/tri_tri_intersection_test.hpp`:

~~~
[100%] Building CXX object test/integration/CMakeFiles/INTEGRATION_heightmap.dir/heightmap.cc.o
In file included from ../gazebo/test/integration/heightmap.cc:19:
In file included from ../gazebo/build/gazebo/rendering/rendering.hh:5:
In file included from ../gazebo/gazebo/rendering/Camera.hh:46:
In file included from ../gazebo/gazebo/rendering/ogre_gazebo.h:26:
In file included from /usr/local/include/OGRE/OgreMatrix4.h:35:
/usr/local/include/OGRE/OgreMatrix3.h:266:27: error: expected member name or ';' after declaration specifiers
        static const Real EPSILON;
        ~~~~~~~~~~~~~~~~~ ^
/usr/local/include/dart/collision/fcl/tri_tri_intersection_test.hpp:45:17: note: expanded from macro 'EPSILON'
#define EPSILON 0.000001
                ^
1 error generated.
~~~

> If you have trouble in `cmake` or `make`, it would be very helpful to build DART with verbose option:
> `$ cmake .. -DDART_VERBOSE=On`
> `$ make VERBOSE=1`
> and then dump the results into a [gist](https://gist.github.com/) and share the link to gist here.

#### Steps to Reproduce
> Please provide detailed steps for reproducing the issue.
1. compile [gazebo pull request 2956 (DART heightmaps)](https://bitbucket.org/osrf/gazebo/pull-requests/2956/dart-heightmap-with-bullet-and-ode/diff) on macOS
1. put `#include <dart/dart.hpp>` at top of gazebo's `test/integration/heightmap.cc`
1. try to build again

#### Code to Reproduce
> Please remember that with a working code it's easier to reproduce the bug and it's much faster to fix it.
> It would be great to fork this repo and create a regression test using [this template](https://github.com/dartsim/dart/blob/master/unittests/regression/test_Issue000Template.cpp).

[gazebo pull request 2956 (DART heightmaps)](https://bitbucket.org/osrf/gazebo/pull-requests/2956/dart-heightmap-with-bullet-and-ode/diff)

#### Discussion

I don't think this is really the fault of dart; I more blame Apple and Ogre for not following the convention of all-caps symbols if-and-only-if it's a `#define` macro. However, exporting lots of `#define` statements in your public API is not an ideal practice either because it can lead to this type of situation.  I don't know that much about `dart/collision/fcl/tri_tri_intersection_test.hpp`; is it intended to be part of the public API? If not, then you could deprecate it and make it a private header.

I'll probably add an `#undef nil` to the gazebo code to workaround this, but I thought I'd mention it here as an FYI that other code is not playing well with `tri_tri_intersection_test.hpp`.

**Comment by mxgrey** [comment]
I don't think this specific header was ever meant to be part of the public API. DART has just historically not made a hard distinction between public and internal API headers to make maintenance easier/simpler.

I think it's probably getting included because of an `#include <dart/collision/fcl/fcl.hpp>` somewhere, which carelessly globs in `dart/collision/fcl/tri_tri_intersection_test.hpp`. A quick solution on the Gazebo side is to not include any master headers and instead only include the specific header that's needed (probably `dart/collision/fcl/FCLCollisionDetector.hpp`).

In the long term, I would be in favor of restructuring dartsim to be more conscientious of which headers are publicly installed versus meant for internal use, but that would probably be a time-consuming refactoring, and I don't know when we would get to it.

Although a quicker fix for dartsim that doesn't involve a large audit/restructure might be to have a blacklist for which headers can globbed into the master headers, and whenever an issue like this pops up we can add the offending header to the blacklist.

**Comment by scpeters** [comment]
Yeah, gazebo is not very selective about its dart headers; it's just `dart/dart.hpp`:

* https://bitbucket.org/osrf/gazebo/src/f4e51e8f64001beda70615ab8982a7d5a4e65b18/gazebo/physics/dart/dart_inc.h?at=default&fileviewer=file-view-default#dart_inc.h-24

Maybe it would be worth our while to be more explicit about which headers we put in `dart_inc.h` file.
